targets/TARGET_STM/can_api.c

```diff
@@ -190,19 +192,42 @@ int can_frequency(can_t *obj, int f)
     int pclk = HAL_RCC_GetPCLK1Freq();
     int btr = can_speed(pclk, (unsigned int)f, 1);
     CAN_TypeDef *can = (CAN_TypeDef *)(obj->can);
+    uint32_t tickstart = 0;
+    int status = 1;
 
     if (btr > 0) {
         can->MCR |= CAN_MCR_INRQ ;
+        /* Get tick */
+        tickstart = HAL_GetTick();
         while ((can->MSR & CAN_MSR_INAK) != CAN_MSR_INAK) {
+            if((HAL_GetTick() - tickstart ) > 2)
+            {
+                status = 0;
+                break;
+            }
         }
-        can->BTR = btr;
-        can->MCR &= ~(uint32_t)CAN_MCR_INRQ;
-        while ((can->MSR & CAN_MSR_INAK) == CAN_MSR_INAK) {
+        if (status != 0) {
+            can->BTR = btr;
+            can->MCR &= ~(uint32_t)CAN_MCR_INRQ;
+            /* Get tick */
+            tickstart = HAL_GetTick();
+            while ((can->MSR & CAN_MSR_INAK) == CAN_MSR_INAK) {
+                if((HAL_GetTick() - tickstart ) > 2)
+                {
+                    status = 0;
+                    break;
+                }
+            }
+            if (status ==0) {
+                error("can ESR  0x%04x.%04x + timeout status %d", (can->ESR&0XFFFF0000)>>16, (can->ESR&0XFFFF), status);
```

Comment:
Would add spaces around `&` and `>>`, also use lowercase `0x` prefix instead of `0X`.